### ISSUE TITLE
Change virtual destructors to be non-inline

### ISSUE THREAD

**Issue by jwnimmer-tri** [RobotLocomotion/drake#22015]
Our style guide [outlaws](https://drake.mit.edu/styleguide/cppguide.html#Inline_Functions) inline virtual destructors, for the very good reason that leaving them inline sets ODR traps for their `type_info` weak objects (among many other important reasons).

In #22010, I cleaned up a few instances that users complained about, but we should circle back and clean all of them up.

As of my current checkout 5f766089, I see 115 instances we need to fix.  Here's the find command:

```
cd drake
find . -name examples -prune -o -name test -prune -o -name test_utilities -prune -o -name '*.h' -exec grep --color -nHE --null -e '(virt.*~.*\(\).*[={]|*~.*\(\).*(over|final).*[={])' \{\} +
```

By directory, that is:
```
     14 ./systems/analysis
     13 ./systems/framework
     12 ./multibody/inverse_kinematics
      8 ./multibody/optimization
      5 ./systems/primitives
      5 ./solvers
      4 ./planning/trajectory_optimization
      4 ./geometry/optimization
      3 ./multibody/contact_solvers
... etc ...
```

Note that open-ended templates (of which we have very few, e.g., `Value<T>`) should *not* be changed.

**Comment by jwnimmer-tri** [comment]
New report as of today:

```
     12 ./multibody/inverse_kinematics
      8 ./multibody/optimization
      7 ./systems/framework
      3 ./multibody/contact_solvers
      2 ./multibody/fem
      2 ./geometry/optimization/dev
      1 ./systems/lcm
      1 ./systems/controllers
      1 ./multibody/triangle_quadrature
      1 ./multibody/parsing
      1 ./multibody/benchmarks/chaotic_babyboot/MG
      1 ./common
```

**Comment by j4yyousi** [comment]
Hi @jwnimmer-tri, why having inline virtual destructors is outlawed? How can they lead to issues? 
My understanding is that a `type_info` object for a polymorphic `class X` is created in each translation unit that contains `class X`. Then the linker combines the equivalent `type_info` objects. Also, classes can’t be redefined in the same translation unit since header files have include guards. Thus a program is well-formed (i.e. builds successfully) in both cases.

**Comment by jwnimmer-tri** [comment]
> ... then the linker combines the equivalent type_info objects ...

Not necessarily.  If there are multiple `*.so` files that all contain the `type_info` (e.g., in our Python bindings), they might not be combined.  That's why we aim to emit just one copy of the `type_info`, so its uniqueness is guaranteed. We build with `--copt=-fno-assume-unique-vtables` to help avoid that trap, but not everyone does.

It's also simply wasteful to emit the vtable (and all of the virtual methods!) into every translation unit, and then let the linker throw away most of them.  Like all other inline methods, the multiple emitted copies bloats the build.  Virtual destructors are just more bloat than most cases since they are usually the first virtual in a class which means it's the "key function" that emits the vtable and everything _it_ needs.

**Comment by j4yyousi** [comment]
> Virtual destructors ... are usually the first virtual in a class which means it's the "key function" that emits the vtable and everything it needs.

I see, so the vtable and `type_info` for a class are emitted in the translation unit that defines the first non-inline virtual function (usually the destructor).

**Comment by jwnimmer-tri** [comment]
I think it's actually the TU that defines the first virtual function, whether inline or not.

**Comment by j4yyousi** [comment]
Hi, are the following template classes 
`./systems/framework/event.h425:  virtual ~Event() = default;`
`./systems/lcm/serializer.h61:  ~Serializer() override = default;`
considered open-ended templates like `Value<T>` and should not be changed since they are not instantiated only on default scalars with a macro like `DRAKE_DEFINE_CLASS_TEMPLATE_INSTANTIATIONS_ON_DEFAULT_SCALARS`?

**Comment by jwnimmer-tri** [comment]
The `Event<T>` is a closed set (so should not be inline):
https://github.com/RobotLocomotion/drake/blob/acf6950da41bb73f1b6c5293b9d852beee98aecc/systems/framework/event.h#L777-L778

The `Serializer<LcmMessage>` is an open set, so must remain inline.

**Comment by j4yyousi** [comment]
Hi, should my last PR close this issue? Since the four remaining instances should be ignored:

1. `./systems/lcm/serializer.h61: ~Serializer() override = default;` is an open-ended template
2. `./multibody/benchmarks/chaotic_babyboot/MG/MG_chaotic_babyboot_auto_generated.h35: virtual ~MGIntegrator() {};` is in a generated file that should not be edited.
3. `./multibody/fem/force_density_field_base.h38: virtual ~ForceDensityFieldBase() = 0;` destructor already defined in `drake/multibody/fem/force_density_field_base.c`
4. `./common/value.h243: ~Value() override = default;` is an open-ended template

**Comment by jwnimmer-tri** [comment]
I see a few more, but after these it would be finished, yes:

```
./common/test_utilities/diagnostic_policy_test_base.h:39:  ~DiagnosticPolicyTestBase() { FlushDiagnostics(); }
./examples/planar_gripper/planar_manipuland_lcm.h:45:  ~PlanarManipulandStatusDecoder() {}
./examples/planar_gripper/gripper_brick_planning_constraint_helper.h:152:  ~FingerNoSlidingConstraint() override {}
./examples/planar_gripper/gripper_brick_planning_constraint_helper.h:190:  ~FingerNoSlidingFromFixedPostureConstraint() override {} 
./examples/rod2d/rod2d.h:159:  ~Rod2D() override {}
./examples/scene_graph/solar_system.h:107:  ~SolarSystem() override = default;
./systems/analysis/test_utilities/controlled_spring_mass_system.h:38:  ~PidControlledSpringMassSystem() override {}
```

**Comment by j4yyousi** [comment]
The command in the first post ignores instances in `test_utilities` and `examples` directories or is that irrelevant?
`find . -name examples -prune -o -name test -prune -o -name test_utilities -prune -o -name '*.h' -exec grep --color -nHE --null -e '(virt.*~.*\(\).*[={]|*~.*\(\).*(over|final).*[={])' \{\} +`

**Comment by jwnimmer-tri** [comment]
They are less critical but we might as well do them, too.  It still affects compilation time, even if our release binaries aren't affected.
